### PR TITLE
MRI-NUFFT operator plugin

### LINKED ISSUE #721: [NewOperator] Add bindings with mri-nufft for non-Cartesian MRI applications
### Environment (optional)

_No response_

### Describe the feature

Bindings with mri-nufft for non-Cartesian MRI 

### Motivation

Non-Cartesian MRI applications (including accelerated MRI and MRSI imaging) support require a non-cartesian MRI operator. 

### Alternatives considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by chaithyagr**
Thank you for contributing to DeepInverse!

Please refer to our [contributing guidelines](https://deepinv.github.io/deepinv/contributing.html) for full instructions on how to contribute, including writing tests, documentation and code style.

Once the GitHub tests have been approved by a maintainer (only required for first-time contributors), and the `Build Docs` GitHub action has run successfully, you can download the documentation as a zip file from the [Actions page](https://github.com/deepinv/deepinv/actions/workflows/documentation.yml). Look for the workflow run corresponding to your pull request.

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: deepinv/datasets/__init__.py [hunk 1]
```diff
@@ -4,11 +4,12 @@
 from .urban100 import Urban100HR
 from .set14 import Set14HR
 from .cbsd68 import CBSD68
-from .fastmri import (
+from .mri.fastmri import (
```
Comment (Andrewwango) [inline deepinv/datasets/__init__.py:7]: CMRxRecon should also go in mri folder

File: deepinv/datasets/__init__.py [hunk 2]
```diff
@@ -4,11 +4,12 @@
...
     FastMRISliceDataset,
     SimpleFastMRISliceDataset,
     MRISliceTransform,
 )
+from .mri.calgary import Calgary3DBrainMRIDataset, CalgaryDataTransformer
```
Comment (Andrewwango) [inline deepinv/datasets/__init__.py:12]: I think `CalgaryDataTransform` would be better to be similar to fastMRI `MRISliceTransform`

File: deepinv/datasets/mri/brainweb.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+r"""A PyTorch dataset of 20 Normal Brains based on Brainweb"""
```
Comment (Andrewwango) [inline deepinv/datasets/mri/brainweb.py:1]: We have a few examples in the library that load brainweb data from our huggingface ([vol](https://huggingface.co/datasets/deepinv/images/blob/main/brainweb_t1_ICBM_1mm_subject_0.npy), [slice](https://huggingface.co/datasets/deepinv/images/blob/main/brainweb_t1_ICBM_1mm_subject_0_slice_0.npy)). For example https://deepinv.github.io/deepinv/auto_examples/physics/demo_mri_tour.html#explore-3d-mri https://deepinv.github.io/deepinv/auto_examples/physics/demo_microscopy_3d.html https://deepinv.github.io/deepinv/auto_examples/optimization/demo_3D_wavelets.html

Given that you are proposing a brainweb dataset here, it would be great to a) have an option in BrainWeb to download a "dummy image" for quick testing (to avoid having `bwdl` as a dependency) and b) change these examples to use BrainWeb dataset with `download_dummy=True`

File: deepinv/datasets/mri/brainweb.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+from deepinv.datasets.base import ImageDataset
+import brainweb_dl as bwdl
```
Comment (Andrewwango) [inline deepinv/datasets/mri/brainweb.py:4]: this will have to be an optional dependency, i.e. only gets imported when `download=True` in the dataset

File: deepinv/physics/mri.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 
 from deepinv.physics.forward import DecomposablePhysics, LinearPhysics
 from deepinv.utils.mixins import MRIMixin, TimeMixin
+from mrinufft import get_operator
```
Comment (Andrewwango) [inline deepinv/physics/mri.py]: Best to defer importing to inside the physics init

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Some initial comments, thank you @chaithyagr ! Please add some tests for MRI-NUFFT (you can see the astra tests as an example)

File: examples/physics/demo_non_cartesian_mri_tour.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
+r"""
+Tour of Non-Cartesian MRI functionality in DeepInverse
```
Comment (Andrewwango) [inline examples/physics/demo_non_cartesian_mri_tour.py]: I think the example should be added to the external-libraries folder, since it's not a core deepinv physics, but rather showcasing our external integrations.

**Review by Andrewwango — CHANGES_REQUESTED** [review]

**Comment by chaithyagr** [comment]
IDeally, this depends on #738 and #743 .

File: deepinv/physics/mri.py [hunk 7]
```diff
@@ -618,3 +618,107 @@ def A_adjoint(
...
+    >>> class TorchNUFFT:
+    ...     def __init__(self, E, device):
+    ...         self.E, self.device = E, device
+    ...     def op(self, img_t):
+    ...         y = self.E.op(img_t.detach().cpu().numpy())          # NumPy
+    ...         return torch.as_tensor(y, device=self.device, dtype=torch.complex64)
+    ...     def adj_op(self, y_t):
+    ...         x = self.E.adj_op(y_t.detach().cpu().numpy())
+    ...         return torch.as_tensor(x, device=self.device, dtype=torch.complex64)
+
```
Comment (paquiteau) [inline deepinv/physics/mri.py:699]: MRI-NUFFT is polite and will return torch array if you provide it with torch array. (see  [with_numpy_cupy](https://github.com/mind-inria/mri-nufft/blob/8e522f40380a8ba6b3a124ccd2f8f6ef7e096fdb/src/mrinufft/_array_compat.py#L46) decorator)

File: deepinv/physics/mri.py [hunk 8]
```diff
@@ -618,3 +618,107 @@ def A_adjoint(
...
+    - Data types & devices. This wrapper does not convert between NumPy
+      and PyTorch. If your NUFFT returns NumPy arrays, wrap it so that both
+      ``op`` and ``adj_op`` return torch tensors on the same device and
+      with a complex dtype (e.g., ``torch.complex64``) compatible with the
+      rest of your pipeline.
+
+    - Coil sensitivities. This wrapper does not estimate or apply
+      sensitivity maps. If you need SENSE/ESPIRiT, pre-apply them in your
+      backend (e.g., pass coil-modulated images to ``op`` and perform the
+      conjugate-weighted sum in your reconstruction algorithm).
```
Comment (paquiteau) [inline deepinv/physics/mri.py:661]: i think this should be updated :) Also a reference link to MRI-NUFFT documentation would probably help a lot.

**Review by paquiteau — COMMENTED** [review]
Unsollicitated review (: 

Thank you @chaithyagr  for building bridges between our two libraries

File: deepinv/physics/mri.py [hunk 9]
```diff
@@ -618,3 +618,107 @@ def A_adjoint(
...
+      linear. Ensure ``E.op`` and ``E.adj_op`` do **not** depend on the current
+      iterate (e.g., no data-dependent coil re-weighting inside them).
+
+    - Data types & devices. This wrapper does not convert between NumPy
+      and PyTorch. If your NUFFT returns NumPy arrays, wrap it so that both
+      ``op`` and ``adj_op`` return torch tensors on the same device and
+      with a complex dtype (e.g., ``torch.complex64``) compatible with the
+      rest of your pipeline.
+
+    - Coil sensitivities. This wrapper does not estimate or apply
```
Comment (Andrewwango) [inline deepinv/physics/mri.py:658]: Is there definitenly no way to mirror the current MultiCoilMRI physics which abstracts the entire operator to do `x` (image) -> `y` (multicoil kspace) with parameters `mask` (trajectory in this case) and `maps`? We don't need map estimation (users could just use MultiCoilMRI), just a way of `update`ing the physics on-the-fly or in the constructor like all deepinv physics i.e.:

```python
maps = ...
trajectory = ...
physics = dinv.physics.NonCartesianMRI(..., trajectory=trajectory, maps=maps)
or
physics.update(maps=maps, trajectory=trajectory)
```

Comment (chaithyagr) [reply deepinv/physics/mri.py:658]: the samples and smaps can just be updated with `nufft_op.samples = new_samples` and nufft_op.smaps = new_smaps` and this shall lead to under the hood updates (as `samples` and `smaps` are just properties)

**Comment by chaithyagr** [comment]
Adding here that mri-nufft natively supports passing back deepinv compatible op with which you can run a reconstruction now: https://mind-inria.github.io/mri-nufft/generated/autoexamples/extras/example_recon.html
Maybe we close this to avoid overburfening this repository? Comments @Andrewwango ?

**Comment by Andrewwango** [comment]
Closing in favour of https://github.com/deepinv/deepinv/pull/1381
